doc/developer/design/20210728_error_handling_surfacing_connector_failures.md

```diff
@@ -0,0 +1,194 @@
+   execution must change the state, instead of just logging error messages.
+ - Add "status" columns to `mz_connector` that indicates the status of a
+   connector. These would surface the lifecycle state mentioned above.
+ - Add a new view `mz_connector_errors` that surface errors that are only
+   reported in the log so far.
+ - Maybe log state transitions in the above new views or add additional views
+   for that.
+ - Failures in a connector should never bring down Materialize. Right now this
+   can happen when re-starting. For those cases we should instead set the
+   lifecycle status accordingly and continue. It's somewhat tricky because this
+   happens when restoring the catalog and looks to the sink like it's just
+   being created. Normally, when creating a sink, failures should cancel the
+   statement and not add the sink to the catalog but we cannot do this here.
+
+#### Dealing with transitive failures
+
+What should be the state of a sink when one of the sources that it depends on
+fails. I think it should also be put in a failed state, with
+`mz_connector_errors` indicating that it was failed because one of its inputs
+has failed.
+
+#### Musing
+
+ - Add something like `RESTART SOURCE` or `RESTART SINK` that can get a
+   source/sink out of a wedged state. This is potentially complicated because
+   restarting sources would also require restarting dependent sinks.
+ - When we add *one-shot sinks* later, we can add a state `FINISHED` that would
+   indicate finished sinks.
+ - Keep entries of dropped/failed/whatever connector in a
+   `mz_connector_history` view. Or keep dropped connector in `mz_connector` for
+   a while.
+
+## Alternatives
+
+Instead of making `timeout`/`num retries` globally configurable, we could leave
+the situation as is and introduce per-connector configuration when needed. The
+other parts of the proposal would be left in place, and the way for users to
+deal with problems would be to `DROP` and re-create a wedged view or sinks.
+
+I think this is a viable alternative because some SDKs, for example Kafka, have
+infinite retries built in for certain operations. For example, our source will
+currently sit and wait indefinitely until the Kafka brokers come up again. (See
+also the open question below.). Only for fatal errors would we mark the
+connector as failed and surface the error in the system views.
+
+The important part of the proposal is specifying (and fixing) how Materialize
+reacts to errors in different stages and how we surface those errors.
+
+## Open Questions
+
+Some SDKs, for example `rdkafka`, will happily sit and not report any errors
+when brokers can't be reached from the consumer. Once the broker is available
+again they will start consuming again. At least that's the case in our setup,
+with split consumer keys. And the consumer will log errors still.
+
+We can be fine with that, or also try and cover this with a timeout and try to
```

Comment:
This has always irked me about librdkafka. I think we'd be well served by finding some way to surface a transient "degraded" state while librdkafka is spewing errors.

Reply:
:+1: Once we have the machinery in place to track connector state this should be doable. **If** we can coax it out of librdkafka...